### PR TITLE
bpo-42035: Add a PyType_GetQualName() to get type's qualified name.

### PR DESCRIPTION
**Pull request by shihai1991**
https://bugs.python.org/issue42035

### CODE REVIEW

**Comment by shihai1991** [comment]
@encukou Hi, Petr. Would you mind to take a look :)

File: Modules/_testcapimodule.c [hunk 1]
```diff
@@ -1159,6 +1159,39 @@ test_get_type_name(PyObject *self, PyObject *Py_UNUSED(ignored))
...
+    assert(strcmp(PyUnicode_AsUTF8(tp_qualname), "SimpleNamespace") == 0);
+    Py_DECREF(tp_qualname);
+
+    PyObject *HeapTypeNameType = PyType_FromSpec(&HeapTypeNameType_Spec);
+    if (HeapTypeNameType == NULL) {
+        Py_RETURN_NONE;
+    }
+    PyObject *spec_name = PyUnicode_FromString(HeapTypeNameType_Spec.name);
+    if (PyObject_SetAttrString(HeapTypeNameType,
+                               "__qualname__", spec_name) < 0) {
```
Comment (encukou) [inline Modules/_testcapimodule.c:1186]: This is an interesting test.
Would it make sense to test the *original* qualname first, then assign a different one and test that `PyType_GetQualName` returns the replacement?

Comment (shihai1991) [reply Modules/_testcapimodule.c:1186]: Haha. I try to get some qual names in _testcapimodule. But I can get the qual name with prefix.
The reason is `res->ht_qualname = res->ht_name;` in https://github.com/python/cpython/blob/main/Objects/typeobject.c#L3439.
I don't know the reason why the qual name need to reomve the prefix too.
So I set a qual name with prefix to test this C API, somthing like a mock behavior~

Comment (encukou) [reply Modules/_testcapimodule.c:1186]: Right, C-API classes don't have distinct qualnames.
IMO, the best thing to do here would be:
- Test the *original* qualname (same as the short name)
- Change the qualname
- Then test the new qualname.

Ideally, the same thing would be tested for `PyType_GetName` as well – it's not that different from this function.

Comment (shihai1991) [reply Modules/_testcapimodule.c:1186]: > Right, C-API classes don't have distinct qualnames.
> IMO, the best thing to do here would be:
> 
> * Test the _original_ qualname (same as the short name)
> * Change the qualname
> * Then test the new qualname.

Make sense. How about this updated PR :)

> Ideally, the same thing would be tested for `PyType_GetName` as well – it's not that different from this function.

Got it. I will update enhance this test case later.

**Comment by encukou** [comment]
Looks good, thanks!

**Comment by shihai1991** [comment]
> Looks good, thanks!

Thanks a million, petr :)

**Pull request merged**